### PR TITLE
Automate resource processor debugging

### LINKED ISSUE #641: [BUG] Permissions missing for running Docker in Codespaces
When trying to execute docker commands in Codespaces the following error appers

vscode ➜ /workspaces/AzureTRE (develop) $ docker version
Client: Docker Engine - Community
 Version:           20.10.8
 API version:       1.41
 Go version:        go1.16.6
 Git commit:        3967b7d
 Built:             Fri Jul 30 19:54:02 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/version": dial unix /var/run/docker.sock: connect: permission denied

To fix it run:

source ./devops/scripts/set_docker_sock_permission.sh

### LINKED ISSUE #1401: Automate set-up of resource processor local debugging
**Is your feature request related to a problem? Please describe.**
Currently there are a lot of manual steps to set up local debugging for the resource processor, whereas the api local debugging is mostly automated,

**Describe the solution you'd like**
We should bring in line the Api debugging set up to encompass the resource processor and consolidate the separate .env file in the `./resource_processor` directory into the central `tre.env` env vars from the Terraform output

**Additional context**
Some of this pre-work has been done in the custom actions PR to output required vars from terraform to `tre.env`

### PR DESCRIPTION
**Pull request by jjgriff93**
# PR for issue

## What is being addressed

This PR adds automation of environment set-up for debugging the resource provider locally. It also tidies up the Resource Processor, including how it uses env vars as config and moving logic to smaller and more manageable files.

## How is this addressed

- Added logic to the `make setup-local-debug` script to create Service Principal for resource processor testing, add service bus permissions to it and export the appId and client secret to env file to be consumed by debug profile
- Simplified config intialisation in RP so config can be reused as a global var instead of passed around every function call
- Moved some RP logic from `runner.py` to smaller modules
- Updated docs
- Fixed a Docker socket permissions error (bug #641) by calling script in devcontainer postCreate

Closes #1401 and also fixes #641

### CODE REVIEW

**Review by tamirkamara — CHANGES_REQUESTED** [review]
It's difficult to assess the changes in the RP since all files are especially new.
Have you made any changes on the logic?
How did you test this?

File: devops/scripts/setup_local_debugging.sh [hunk 1]
```diff
@@ -6,6 +6,7 @@ set -e
 : ${SERVICE_BUS_RESOURCE_ID?"Check SERVICE_BUS_RESOURCE_ID is defined in ./templates/core/private.env"}
```
[the remark this answers is not in the document]
Comment (jjgriff93) [reply devops/scripts/setup_local_debugging.sh:6]: Agreed, this was where it resided when I edited it but makes sense to move

[the remark this answers is not in the document]
Comment (jjgriff93) [reply devops/scripts/setup_local_debugging.sh:6]: Moved

File: devops/scripts/setup_local_debugging.sh [hunk 2]
```diff
@@ -41,3 +42,33 @@ az role assignment create \
     --role "Contributor" \
     --assignee ${LOGGED_IN_OBJECT_ID} \
     --scope ${STATE_STORE_RESOURCE_ID}
+
+# Configure SP for local resource processor debugging (Porter can't use local creds)
+echo "Configuring Service Principal for Resource Processor debugging..."
+RP_TESTING_SP=$(az ad sp create-for-rbac --name ResourceProcessorTesting --role Owner --scopes /subscriptions/${AZURE_SUBSCRIPTION_ID} -o json)
+RP_TESTING_SP_APP_ID=$(echo ${RP_TESTING_SP} | jq .appId | tr -d '"')
+RP_TESTING_SP_PASSWORD=$(echo ${RP_TESTING_SP} | jq .password | tr -d '"')
```
[the remark this answers is not in the document]
Comment (jjgriff93) [reply devops/scripts/setup_local_debugging.sh]: Ah yes, thanks!

File: devops/scripts/setup_local_debugging.sh [hunk 3]
```diff
@@ -41,3 +42,33 @@ az role assignment create \
...
+    --assignee ${RP_TESTING_SP_APP_ID} \
+    --scope /subscriptions/${AZURE_SUBSCRIPTION_ID}
+
+# Write the appId and secret to the private.env file which is used for RP debugging
+# First check if the env vars are there already and delete them
+sudo sed -i '/ARM_CLIENT_ID/d' ./templates/core/private.env
+sudo sed -i '/ARM_CLIENT_SECRET/d' ./templates/core/private.env
+
+# Append them to the TRE file so that the Resource Processor can use them
+sudo tee -a ./templates/core/private.env <<EOF
```
[the remark this answers is not in the document]
Comment (jjgriff93) [reply devops/scripts/setup_local_debugging.sh]: Removed

**Comment by jjgriff93** [comment]
> 

@tamirkamara No changes to the logic whatsoever and have tested locally that the resource processor still functions as expected and all e2e tests ran successfully

**Review by tamirkamara — APPROVED** [review]

**Pull request merged**